Workbook package

ChapC opened the pull request:
### Category

- [ ] Bug fix?
- [x] New feature?
- [ ] New sample?
- [ ] Documentation update?

### Related Issues

Mentioned in #3016 

### What's in this Pull Request?
Initial draft of Excel workbook support. 
This first commit includes
- workbook session creation
- listing/getting tables
- listing/getting rows

I'll implement more endpoints as I go, but have a couple of notes/questions I thought I'd post here first.

1. Your contributing guidelines mention I should skip writing tests for endpoints that don't support app auth. According to the docs, that's all of the workbook endpoints. Is there any way I can write tests we can use for these endpoints?
2. In my testing, I wasn't able to use workbook operations on items opened by path. These failed with "Could not obtain a WAC token". When I use getItemById, it works. I wrote this off as a Graph API bug, but haven't tested extensively.
3. My current implementation of workbook sessions involves a `workbook` property on DriveItem that leads to a regular, sessionless IWorkbook and an async method `getWorkbookSession` that resolves to an IWorkbookWithSession. The latter is an extension of IWorkbook that adds session refresh/close methods, and uses the InjectHeaders behaviour to add the session id header. Thoughts on this approach?
4. I've added a prefer-async behavior to the graph folder. This is to handle the long-running operation pattern that a few of the workbook endpoints support. Again, I'm not sure if this is the ideal approach, but it's what I've got at the moment.

Welcoming all feedback. Thanks!

ChapC committed:
Initial workbook package layout

ChapC committed:
Add .js extension to import

ChapC committed:
TableColumn type

patrick-rodgers commented on packages/graph/behaviors/prefer-async.ts:
```diff
@@ -0,0 +1,39 @@
+                let succeeded = false;
+                const statusQuery = GraphQueryable(instance, `operations/${opId}`);
+                for (let i = 0; i < maxPolls; i++) {
+                    await new Promise(resolve => setTimeout(resolve, pollIntervalMs));
```
We have [delay()](https://github.com/pnp/pnpjs/blob/version-4/packages/core/util.ts#L180) in core you can use for this very thing.

patrick-rodgers commented on packages/graph/behaviors/prefer-async.ts:
```diff
@@ -0,0 +1,39 @@
+                for (let i = 0; i < maxPolls; i++) {
+                    await new Promise(resolve => setTimeout(resolve, pollIntervalMs));
+
+                    const status = await graphGet<RichLongRunningOperation>(statusQuery);
```
For GET you should be able to invoke the queryable itself `await statusQuery<>()`

patrick-rodgers commented on packages/graph/behaviors/prefer-async.ts:
```diff
@@ -0,0 +1,39 @@
+        instance.using(InjectHeaders({ "Prefer": "respond-async" }));
+
+        instance.on.parse(async function (url, response, result) {
+            if (response.status === 202) {
```
What happens if you get a different response? We have some standard logic [wrapped up here](https://github.com/pnp/pnpjs/blob/version-4/packages/queryable/behaviors/parsers.ts#L100) to check for errors. Used [here](https://github.com/pnp/pnpjs/blob/version-4/packages/queryable/behaviors/parsers.ts#L7) and [here](https://github.com/pnp/pnpjs/blob/version-4/packages/queryable/behaviors/parsers.ts#L23) for example.

patrick-rodgers commented on packages/graph/workbooks/session.ts:
```diff
@@ -0,0 +1,20 @@
+    WorkbookSessionInfo
+} from "@microsoft/microsoft-graph-types";
+
+export async function getWorkbookSession(this: _DriveItem, persistChanges: boolean): Promise<IWorkbookWithSession> {
```
If the only place this will really get used is in the property in the other file you can code it there directly. i.e. remove the export here and put the code there. But if this is a function that will be used on its own this makes sense.

patrick-rodgers reviewed (commented):
Really nice work! You did a great job figuring out and integrating with the library. I left a few comments for you to consider. Excited for your contribution!!

patrick-rodgers commented:
**1. Your contributing guidelines mention I should skip writing tests for endpoints that don't support app auth. According to the docs, that's all of the workbook endpoints. Is there any way I can write tests we can use for these endpoints?**

Not really. You can test them locally. We are working on test recording to allow us to record delegated-auth tests and replay them to validate.

**2. In my testing, I wasn't able to use workbook operations on items opened by path. These failed with "Could not obtain a WAC token". When I use getItemById, it works. I wrote this off as a Graph API bug, but haven't tested extensively.**

Don't know, WAC is the office protocol, perhaps that endpoint requires something special.

**3. My current implementation of workbook sessions involves a `workbook` property on DriveItem that leads to a regular, sessionless IWorkbook and an async method `getWorkbookSession` that resolves to an IWorkbookWithSession. The latter is an extension of IWorkbook that adds session refresh/close methods, and uses the InjectHeaders behaviour to add the session id header. Thoughts on this approach?**

I am not deeply familiar with this API, what is the difference between sessionless and with session? Do we need to expose both?

**4. I've added a prefer-async behavior to the graph folder. This is to handle the long-running operation pattern that a few of the workbook endpoints support. Again, I'm not sure if this is the ideal approach, but it's what I've got at the moment.**

This is pretty slick, I left a few notes inline. This isn't something we built-in for v4 launch, but I can see a ton of value to including it.

Again, nice work!

ChapC committed:
Added getItemAt decorator
A few of the collections from the workbook endpoints support
getItemAt(index). I've implemented this in the same pattern as
the getById decorator. Currently I've only added it to the rows
collection.

ChapC committed:
Made suggested changes to prefer-async
- Replaced manual setTimeout with delay()
- Directly invoked statusQuery rather than calling graphGet
- Wrapped handler with standard library error checking via
  parseBinderWithErrorCheck

ChapC committed:
Moved getWorkbookSession into driveitem

ChapC committed:
Manually call errorCheck
Oops! I didn't read parseBinderWithErrorCheck properly and missed that
it runs the impl function on all HTTP successes. I only want
prefer-async to trigger on 202 Accepted. Updated my behaviour to
manually add an errorCheck parse handler before my own.

bcameron1231 commented:
HI @ChapC is this ready? We see your changes, we were just waiting on anymore feedback from you.

ChapC committed:
Added additional Range-related endpoints
Endpoints for getting the Range contained by tables, rows, columns
and worksheets. Plus added the 'clearFilter' endpoint to Table.

ChapC commented:
Hey @bcameron1231, I've just added a few more of the endpoints I've tested recently. I was hoping to cover much more of the workbook functionality, but I've ended up with much less time than I thought I would have on this. Apologies!

So far, of [all the workbook sections](https://learn.microsoft.com/en-us/graph/api/resources/excel), I've implemented the basic stuff in Tables, Worksheets, Rows/Columns and Ranges. Everything to get/update/delete data in the spreadsheet. I have mocha tests for most of these on my local machine, although they're in a separate project as I had trouble getting it working with this one.

There were a couple quirks I encountered in these endpoints. Most I mentioned previously, but the biggest one I couldn't find a fix for was incompatibility with paging. The async iterator doesn't work for table rows, because as far as I can tell the Graph API simply doesn't return `hasNext`, `count`, etc. for the `workbook/{table name|id}/rows` endpoint. I've replicated this in Graph explorer as well. You can add `top` and `skip`, just fine, but `count` has no effect and the responses never include `@odata.nextLink`.

There's still quite a bit more for full coverage of all endpoints, like charts, running formulas and cell styling. I'm happy to keep chipping away at these features, just won't have a timeframe for you unfortunately. I can get some documentation written up soon if you guys decide you'd like to merge just the stuff I've got working rather than the full set. Just let me know :)

ChapC commented:
Also, to answer the question from @patrick-rodgers regarding session vs sessionless; my understanding is that both work for pretty much every endpoint, but using a session leads to better performance. [It's briefly explained in the docs here](https://learn.microsoft.com/en-us/graph/excel-manage-sessions). To quote them directly:

> The session header is not required for an Excel API to work. However, we recommend that you use the session header to improve performance. 

It also allows you to create non-persistent sessions, in which you can edit the workbook and run calculations without your changes being saved to the file. In sessionless mode, your changes are always immediately saved to the file.

patrick-rodgers commented:
Sounds good @ChapC! No worries, we're all busy so appreciate any time you have to contribute 😀 Let us know once you're ready for a full review. Thank you!!

ChapC committed:
Range formatting
Added endpoints for Range formatting. That includes borders, fill,
font, row/column sizing, text alignment and protection.

ChapC committed:
Finished table endpoints
Added a few more endpoints (sorting, various range getters) to fully
cover table manipulation.

ChapC committed:
Column filtering

ChapC committed:
Range delete + first few sub-range selectors

ChapC committed:
All other range methods

ChapC committed:
RangeSort and RangeView

ChapC committed:
Worksheet protection & pivot table refresh

ChapC committed:
NamedItem

ChapC committed:
Icon (kind of)
Added the Icon class, but I think the Graph API documentation is wrong.
It's basically useless. See comments for details.

ChapC committed:
Workbook operations

ChapC committed:
Comments

ChapC committed:
Application recalculate

ChapC committed:
Added types to IUpdatable and IAddable
Didn't realise I need to type those. Whoops!

ChapC commented:
After a few more bursts of work, I've finished the majority of endpoints and now have something I think could be worth merging! The last group of endpoints still missing are the Chart-related ones (chart creation, data selection, styling). For now, I think I'll leave those ones as a neat package for another PR down the road (either from myself or someone else who's keen).

The other thing I still need to write is documentation. As you can see from the [MS docs](https://learn.microsoft.com/en-us/graph/api/resources/excel), there are quite a few sections to cover. I'm planning on having most of them just point straight to the docs and adding a few examples for common tasks/quirks.

There are also a few places where the official docs appear to be incorrect. I've noted these as comments in the code where applicable. The biggest of those quirks are the ones I mentioned earlier in this thread: async paging doesn't seem to work and you must open a workbook on an DriveItem ID rather than path. If it's something that's out of our hands, I'll just leave a note in the documentation.

Let me know what you think!

patrick-rodgers commented:
@ChapC  - this is awesome! On the docs - you don't need to go too crazy - really just pointers on how to use THIS library. No need to document the full feature, that's the product's job. You can check out our other docs for how we've approached it - very minimal just to get folks started.

One small nit we found doing another review is the `getItemAt` decorator seems to only be Excel specific? If so, let's move that into the types.ts for excel.

Once you have the docs this looks really solid, excited to get it merged!! Thank you for taking the time to contribute 🙂

ChapC committed:
Move excel-specific decorators into types.ts

juliemturner commented:
Hi @ChapC -- Looks like you addressed the modifications we requested, are we ready to merge this in?

ChapC commented:
Thanks for checking in on this Julie. Happy new year!

Feature-wise we're all set, I just haven't had time to write the docs yet.

I'm hoping to have a few hours to spend on this in the next couple of weeks. Shouldn't take too long as I'm mostly pulling example code snippets from my local tests.

juliemturner commented:
Sounds good, thanks for the update.

ChapC committed:
Hide getByIndex on rows

ChapC committed:
Expose getItemAt on columns

ChapC committed:
Formatting

ChapC committed:
Documentation for named tables

ChapC committed:
Docs for using ranges

ChapC commented:
I think I'm all wrapped up on the docs ✅

juliemturner commented:
@ChapC  - Awesome, should we move it in then, it's still marked draft so I just want to make sure you're ready.

ChapC commented:
Oops, yes sorry! Marked as ready.

juliemturner committed:
Update package-lock.json

Update to match version in v4 branch.

juliemturner reviewed (dismissed):
Updated package-lock.json to match v4 branch.

juliemturner commented:
@ChapC - Hi. I approved the update (after fixing the outdated package-lock.json -- in the future best to avoid checking in those foundational files) but you apparently have linting errors. I could fix them for you but that would mean that I would end up having to close this pr and merging it in as my own. Let me know if you'd like to update the issues the linter found. You can run your own lint check locally by running `npm run lint`

ChapC committed:
eslint autofix

ChapC commented:
Thanks for the heads up @juliemturner. Just ran eslint fix mode and your lint check seems to be happy on my end ✨
Sorry I missed that step from your guidelines!

juliemturner commented:
No worries, we've all done it a million times too. Thanks for your contribution!
